### PR TITLE
Re-implement api/cli using libuplink (V3)

### PR DESCRIPTION
**Pull request by bryanchriswhite**
(updated description in [new PR](https://github.com/storj/libstorj/pull/460))

### CODE REVIEW

File: src/cli.c [hunk 1]
```diff
@@ -15,18 +15,24 @@
...
-    char *pass;
     char *host;
-    char *mnemonic;
-    char *key;
+    char *apikey;
+    char *enc_access;
+    char *user_passphrase;
 } user_options_t;
 
 #ifndef errno
```
Comment (kaloyan-raev) [inline src/cli.c]: The `HELP_TEXT` a few lines below need an update. At least, we should not have the `register` command anymore.

Comment (kaloyan-raev) [reply src/cli.c]: `CLI_VERSION` should be also updated to something like "libstorj-2.0.0-v3-beta"

File: src/cli.c [hunk 2]
```diff
@@ -385,17 +392,36 @@ static int get_password_verify(char *prompt, char *password, int count)
...
+    if (!apikey) {
+        apikey = calloc(BUFSIZ, sizeof(char));
+        printf("Project API key: ");
+        get_input(apikey);
+        printf("\n");
+        num_chars = strlen(apikey);
+
+        if (num_chars == 0) {
+            // TODO: add link
+            printf("Project API key required. See <link here>\n");
```
Comment (kaloyan-raev) [inline src/cli.c]: What should we link here?

File: src/cli.c [hunk 3]
```diff
@@ -425,74 +451,56 @@ static int import_keys(user_options_t *options)
...
 
-    if (!pass) {
-        printf("Bridge password: ");
-        pass = calloc(BUFSIZ, sizeof(char));
-        if (!pass) {
-            status = 1;
-            goto clear_variables;
-        }
-        get_password(pass, '*');
+        printf("Enter your serialized encryption key (leave blank to generate new): ");
```
Comment (kaloyan-raev) [inline src/cli.c]: We are expecting a "serialized encryption access" here, not an encryption key, right?

File: src/cli.c [hunk 4]
```diff
@@ -425,74 +451,56 @@ static int import_keys(user_options_t *options)
...
-        if (!mnemonic_input) {
-            status = 1;
-            goto clear_variables;
-        }
+        if (num_chars == 0) {
+            int last_result = -1;
+            for (int i = 0; i < 3; i++) {
+                last_result = get_password_verify("Enter a passphrase to generate "
+                                                  "a new root key for your "
+                                                  "encryption key: ", new_enc_passphrase, 0);
```
Comment (kaloyan-raev) [inline src/cli.c]: IMHO, a better message would be: "Enter a passphrase to create a new serialized encryption access: "

File: src/cli.c [hunk 5]
```diff
@@ -425,74 +451,56 @@ static int import_keys(user_options_t *options)
...
-        } else {
-            mnemonic = calloc(num_chars + 1, sizeof(char));
-            if (!mnemonic) {
-                status = 1;
-                goto clear_variables;
-            }
-            memcpy(mnemonic, mnemonic_input, num_chars * sizeof(char));
+
+//        } else {
+//          TODO: need to mainpulate `enc_access_str`
```
Comment (kaloyan-raev) [inline src/cli.c]: Is this TODO obsolete? Or is there anything else to be done on `enc_access_str`?

File: src/cli.c [hunk 6]
```diff
@@ -700,39 +583,35 @@ static int export_keys(char *host)
...
-        if (storj_decrypt_read_auth(user_file, key, &user, &pass, &mnemonic)) {
+        if (storj_decrypt_read_auth(user_file, user_passphrase, &apikey, &enc_access_str)) {
             printf("Unable to read user file.\n");
             status = 1;
             goto clear_variables;
         }
 
-        printf("Username:\t%s\nPassword:\t%s\nEncryption key:\t%s\n",
-               user, pass, mnemonic);
+        printf("Passphrase:\t%s\nEncryption key:\t%s\n", apikey, enc_access_str);
```
Comment (kaloyan-raev) [inline src/cli.c]: This message should be `"API key:\t%s\nSerialized encryption access:\t%s\n"`

File: src/cli.c [hunk 7]
```diff
@@ -855,696 +729,582 @@ int main(int argc, char **argv)
...
+    if (!apikey) {
+        apikey = calloc(BUFSIZ, sizeof(char));
+        printf("Project API key: ");
+        get_input(apikey);
         printf("\n");
 
-        user_options_t user_opts = {strdup(user), strdup(pass), strdup(host), NULL, NULL};
+        if (strlen(apikey) == 0) {
+            // TODO: add link
+            printf("Project API key required. See <link here>\n");
```
Comment (kaloyan-raev) [inline src/cli.c]: Another `<link here>`.

File: src/cli.c (hunk 2 above)
Comment (bryanchriswhite) [reply src/cli.c]: Docs explaining where to get an apikey or something was my thought.

File: src/cli.c (hunk 3 above)
Comment (bryanchriswhite) [reply src/cli.c]: Correct, my understanding was that going forward the user-facing language for encryption access context whatever was going to be "encryption key"

Comment (kaloyan-raev) [reply src/cli.c]: I am not sure about this... We create encryption access from the encryption key, which we derive from the passphrase. So the encryption access and the encryption key should be different things, although related.

File: src/storj.h [hunk 8]
```diff
@@ -48,80 +46,66 @@ extern "C" {
...
-#define STORJ_REPORT_SUCCESS 1000
-#define STORJ_REPORT_FAILURE 1100
-
-// Exchange report messages
-#define STORJ_REPORT_FAILED_INTEGRITY "FAILED_INTEGRITY"
-#define STORJ_REPORT_SHARD_DOWNLOADED "SHARD_DOWNLOADED"
-#define STORJ_REPORT_SHARD_UPLOADED "SHARD_UPLOADED"
-#define STORJ_REPORT_DOWNLOAD_ERROR "DOWNLOAD_ERROR"
-#define STORJ_REPORT_UPLOAD_ERROR "TRANSFER_FAILED"
+#define STORJ_QUEUE_ERROR 4000
```
Comment (kaloyan-raev) [inline src/storj.h]: Why are we changing this from 5000 to 4000?

File: src/storj.h [hunk 9]
```diff
@@ -48,80 +46,66 @@ extern "C" {
...
-#define STORJ_FARMER_REQUEST_ERROR 2000
-#define STORJ_FARMER_TIMEOUT_ERROR 2001
-#define STORJ_FARMER_AUTH_ERROR 2002
-#define STORJ_FARMER_EXHAUSTED_ERROR 2003
-#define STORJ_FARMER_INTEGRITY_ERROR 2004
+// Libuplink error (i.e. check STORJ_LAST_ERROR)
+#define STORJ_LIBUPLINK_ERROR 1000
+
+// Memory related errors
+#define STORJ_MEMORY_ERROR 2000
```
Comment (kaloyan-raev) [inline src/storj.h]: Why are we changing this from 4000 to 2000?

File: src/storj.h [hunk 10]
```diff
@@ -48,80 +46,66 @@ extern "C" {
...
-#define STORJ_BRIDGE_OFFER_ERROR 1014
-
-// Farmer related errors 2000 to 2999
-#define STORJ_FARMER_REQUEST_ERROR 2000
-#define STORJ_FARMER_TIMEOUT_ERROR 2001
-#define STORJ_FARMER_AUTH_ERROR 2002
-#define STORJ_FARMER_EXHAUSTED_ERROR 2003
-#define STORJ_FARMER_INTEGRITY_ERROR 2004
+// Libuplink error (i.e. check STORJ_LAST_ERROR)
+#define STORJ_LIBUPLINK_ERROR 1000
```
Comment (kaloyan-raev) [inline src/storj.h]: This looks to a new error constant. Let's use a number that was not used before.

**Comment by bryanchriswhite** [comment]
So I made a git mistake and as a result this PR needs to [move here](https://github.com/storj/libstorj/pull/460) I guess.

**Pull request merged**